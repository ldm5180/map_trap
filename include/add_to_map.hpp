#ifndef ADD_TO_MAP_HPP
#define ADD_TO_MAP_HPP

#include <algorithm>
#include <benchmark/benchmark.h>
#include <cstdlib>
#include <string>

static const auto add_to_map = [](const auto num_elems, auto &map) {
  // Add elements to map, this sorts automatically.
  for (auto i = 0; i < num_elems; ++i) {
    const auto random_key = std::rand();
    map[random_key] = std::to_string(map.size());
  }
  benchmark::ClobberMemory();
};

static const auto add_to_map_and_key_list = [](const auto num_elems, auto &map,
                                               auto &keys) {
  // Add elements to map, this sorts automatically.
  for (auto i = 0; i < num_elems; ++i) {
    const auto random_key = std::rand();
    map[random_key] = std::to_string(map.size());
    keys.push_back(random_key);
    std::sort(keys.begin(), keys.end());
  }
  benchmark::ClobberMemory();
};

#endif // ADD_TO_MAP_HPP
