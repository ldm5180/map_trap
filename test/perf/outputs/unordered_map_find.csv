Run on (4 X 2594.09 MHz CPU s)
2017-10-28 16:41:55
name,iterations,real_time,cpu_time,time_unit,bytes_per_second,items_per_second,label,error_occurred,error_message
"BM_unordered_map_find/8",1663596,503.544,503.23,ns,,,,,
"BM_unordered_map_find/16",1729584,419.531,419.014,ns,,,,,
"BM_unordered_map_find/32",1761535,418.053,418.028,ns,,,,,
"BM_unordered_map_find/64",1751461,428.523,428.491,ns,,,,,
"BM_unordered_map_find/128",1745629,420.656,420.625,ns,,,,,
"BM_unordered_map_find/256",1000000,518.082,517.749,ns,,,,,
"BM_unordered_map_find/512",1455221,461.785,461.093,ns,,,,,
"BM_unordered_map_find/1024",1567402,560.521,560.141,ns,,,,,
"BM_unordered_map_find/2048",1000000,856.782,852.408,ns,,,,,
