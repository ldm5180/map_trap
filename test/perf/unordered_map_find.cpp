#include "add_to_map.hpp"

#include <benchmark/benchmark.h>
#include <cstdlib>
#include <unordered_map>

using map_t = std::unordered_map<int64_t, int64_t>;

static void BM_unordered_map_find(benchmark::State &state) {
  std::srand(42);
  map_t map{};

  for (auto i = 0; i < state.range(0); ++i) {
    add_to_map(state.range(0), map);
  }

  for (auto _ : state) {
    map_t::mapped_type v{};
    benchmark::DoNotOptimize(v = map[std::rand()]);
  }
}
BENCHMARK(BM_unordered_map_find)->RangeMultiplier(2)->Range(8, 8 << 8);
BENCHMARK_MAIN();
