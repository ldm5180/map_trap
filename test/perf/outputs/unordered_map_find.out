Run on (4 X 2594.09 MHz CPU s)
2017-10-28 16:30:20
name,iterations,real_time,cpu_time,time_unit,bytes_per_second,items_per_second,label,error_occurred,error_message
"BM_unordered_map_find/8",1764961,507.59,507.804,ns,,,,,
"BM_unordered_map_find/16",1744341,433.649,433.351,ns,,,,,
"BM_unordered_map_find/32",1784324,426.472,426.681,ns,,,,,
"BM_unordered_map_find/64",1749196,431.365,431.58,ns,,,,,
"BM_unordered_map_find/128",1729121,436.767,436.164,ns,,,,,
"BM_unordered_map_find/256",1000000,502.567,502.817,ns,,,,,
"BM_unordered_map_find/512",1407214,442.338,442.559,ns,,,,,
"BM_unordered_map_find/1024",1066235,478.677,478.908,ns,,,,,
"BM_unordered_map_find/2048",1000000,871.046,870.825,ns,,,,,
